Add metrics for temporal workflow resets

jdpgrailsdev opened the pull request:
## What
* Add visibility into execution of Temporal workflow reset jobs

## How
* Record custom metrics tracking the reset attempt and success or failure.

## Recommended reading order
1. `OssMetricsRegistry.java`
2. `ConnectionManagerWorkflowImpl.java`

## Additional Details

Per #13773, this PR records metrics for the cases mentioned in that ticket, plus two additional ones:  attempts and cancellations.  The reason for adding attempts is to make it possible to compute rates of success/failure, if necessary.  With it out, you would need to add all of the metrics together to get a total number of reset job attempts.  

This is probably a good first pass, but ultimately, this would be better served by something like DD or OTEL traces, which would allow for dimensions to be set on the value.  The benefit of this approach would be to be able to pivot by additional information (e.g. connector type, etc). 

cc: @davinchia @lmossman

alovew commented:
>> ultimately, this would be better served by something like DD or OTEL traces

I think this actually is using Datadog or OTEL, depending on what is configured by the Metric Client Factory when it's initialized. I think this might not be working though, since the METRIC_CLIENT env variable is required and I don't think those are correctly passed into the worker pods. I think to pass the DD env variables, you'd have to do something like @benmoriceau did in this PR for the stream capable state env variable: https://github.com/airbytehq/airbyte/pull/14936/files#diff-695d34ecf14b05b2caa0a54129d04965bf08679ee2c4a5dc7d97ca8cfa900154

For testing this in dev, you have to use dev (not dev-1, dev-2, etc) since that is the only dev cluster with the datadog environment variables enabled, but right now there's actually a bug in deploying to dev (see slack thread here: https://airbytehq-team.slack.com/archives/C02TXT37Y3Y/p1658529054884449

jdpgrailsdev commented:
> > > ultimately, this would be better served by something like DD or OTEL traces
> 
> I think this actually is using Datadog or OTEL, depending on what is configured by the Metric Client Factory when it's initialized. I think this might not be working though, since the METRIC_CLIENT env variable is required and I don't think those are correctly passed into the worker pods. I think to pass the DD env variables, you'd have to do something like @benmoriceau did in this PR for the stream capable state env variable: https://github.com/airbytehq/airbyte/pull/14936/files#diff-695d34ecf14b05b2caa0a54129d04965bf08679ee2c4a5dc7d97ca8cfa900154
> 
> For testing this in dev, you have to use dev (not dev-1, dev-2, etc) since that is the only dev cluster with the datadog environment variables enabled, but right now there's actually a bug in deploying to dev (see slack thread here: https://airbytehq-team.slack.com/archives/C02TXT37Y3Y/p1658529054884449

@alovew I should have been more clear with my critique:  we are not exposing the ability to set attributes on the metric, which creates some duplication.  Typically, I like to do things like have a single `failure` metric that has an attribute that has the type of failure.  This is to 1) reduce the number of metrics we create and 2) allow us to do all sorts of grouping to understand the types of failures or just sum they all up to get a total count of failures.  

I will deploy and test in dev once we agree upon the metric names, etc.

jdpgrailsdev commented:
> > > > ultimately, this would be better served by something like DD or OTEL traces
> > 
> > 
> > I think this actually is using Datadog or OTEL, depending on what is configured by the Metric Client Factory when it's initialized. I think this might not be working though, since the METRIC_CLIENT env variable is required and I don't think those are correctly passed into the worker pods. I think to pass the DD env variables, you'd have to do something like @benmoriceau did in this PR for the stream capable state env variable: https://github.com/airbytehq/airbyte/pull/14936/files#diff-695d34ecf14b05b2caa0a54129d04965bf08679ee2c4a5dc7d97ca8cfa900154
> > For testing this in dev, you have to use dev (not dev-1, dev-2, etc) since that is the only dev cluster with the datadog environment variables enabled, but right now there's actually a bug in deploying to dev (see slack thread here: https://airbytehq-team.slack.com/archives/C02TXT37Y3Y/p1658529054884449
> 
> @alovew I should have been more clear with my critique: we are not exposing the ability to set attributes on the metric, which creates some duplication. Typically, I like to do things like have a single `failure` metric that has an attribute that has the type of failure. This is to 1) reduce the number of metrics we create and 2) allow us to do all sorts of grouping to understand the types of failures or just sum they all up to get a total count of failures.
> 
> I will deploy and test in dev once we agree upon the metric names, etc.

I stand corrected:  it looks like we do translate the tags into attributes.  I am going to re-work this PR.

jdpgrailsdev commented:
@benmoriceau @alovew @gosusnp @davinchia 

I have amended this PR to include a refactor of the `MetricsClient` so that it properly records attributes.  Before this change, it took in an array of strings/tags and used each tag as both the key and value when generating metric attributes for OTEL metrics.  This defeats the purpose of attributes, which are meant to group metrics.  With this change, we can have one `failure` metric for the reset workflow execution with a different `cause`.  It also gives us the ability to add other attributes if we determine there is some other beneficial grouping.

gosusnp commented on airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:
```diff
@@ -226,13 +235,15 @@ private CancellationScope generateSyncWorkflowRunnable(final ConnectionUpdaterIn
           // do nothing, cancellation handled by cancellationScope
-
+          metricClient.count(OssMetricsRegistry.TEMPORAL_WORKFLOW_RESTART_FAILURE, 1L,
+              new MetricAttribute(RESET_WORKFLOW_FAILURE_ATTRIBUTE_KEY, "CANCELED"));
```
nit: I find the `_ATTRIBUTE_KEY` not super clear, what do you think of `RESET_WORKFLOW_FAILURE_CAUSE`?

jdpgrailsdev replied:
Per @davinchia 's comment above, this is going to move to the `MetricTags` class and match the naming conventions used there.

gosusnp commented on airbyte-metrics/metrics-lib/src/main/java/io/airbyte/metrics/lib/DogStatsDMetricClient.java:
```diff
@@ -83,34 +85,44 @@ public void count(final MetricsRegistry metric, final long amt, final String...
+   * @return An array of tag values.
+   */
+  private String[] toTags(final MetricAttribute... attributes) {
+    return Stream.of(attributes).map(a -> a.value()).collect(Collectors.toList()).toArray(new String[] {});
```
Shouldn't we try to keep the `a.key()` as well?

benmoriceau replied:
I agree, it is confusing to remove it. It feels that it is useless since it is never send to data dog.

davinchia commented on airbyte-metrics/metrics-lib/src/main/java/io/airbyte/metrics/lib/OssMetricsRegistry.java:
```diff
@@ -95,7 +95,19 @@ public enum OssMetricsRegistry implements MetricsRegistry {
+      "overall job runtime - scheduling and execution for all attempts - for jobs that reach terminal states in the last hour. tagged by terminal states."),
+
+  TEMPORAL_WORKFLOW_RESTART_ATTEMPT(MetricEmittingApps.WORKER,
+      "temporal.workflow.restart.attempt",
```
is there a reason to break with the underscore convention?

the original names were formed according to https://docs.datadoghq.com/developers/guide/what-best-practices-are-recommended-for-naming-metrics-and-tags/):

jdpgrailsdev replied:
@davinchia Nope...can change.

davinchia commented on airbyte-metrics/metrics-lib/src/main/java/io/airbyte/metrics/lib/MetricAttribute.java:
```diff
@@ -0,0 +1,10 @@
+package io.airbyte.metrics.lib;
+
+/**
+ * Custom tuple that represents a key/value pair to be included with a metric.
```
thinking out loud, is it worth noting here (or anywhere else) how the different classes consume this?

jdpgrailsdev replied:
@davinchia My gut says to document the override in each client, as any comment we put here will quickly become out of date.  However, I will add that it is up to the client to determine what fields to use from this tuple.

davinchia commented on airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/JobCreationAndStatusUpdateActivityImpl.java:
```diff
@@ -229,7 +233,7 @@ public void attemptFailure(final AttemptFailureInput input) {
       for (final FailureReason reason : failureSummary.getFailures()) {
         MetricClientFactory.getMetricClient().count(OssMetricsRegistry.ATTEMPT_FAILED_BY_FAILURE_ORIGIN, 1,
-            MetricTags.getFailureOrigin(reason.getFailureOrigin()));
+            new MetricAttribute("failure_origin", MetricTags.getFailureOrigin(reason.getFailureOrigin())));
```
we can use the https://github.com/airbytehq/airbyte/blob/master/airbyte-metrics/metrics-lib/src/main/java/io/airbyte/metrics/lib/MetricTags.java#L17 to avoid having small duplicated strings over the codebase.

davinchia commented on airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/JobCreationAndStatusUpdateActivityImpl.java:
```diff
@@ -52,6 +53,8 @@
 @Slf4j
 public class JobCreationAndStatusUpdateActivityImpl implements JobCreationAndStatusUpdateActivity {
 
+  private static final String RELEASE_STAGE_ATTRIBUTE_KEY = "release_stage";
```
nit: i would think about moving this to https://github.com/airbytehq/airbyte/blob/master/airbyte-metrics/metrics-lib/src/main/java/io/airbyte/metrics/lib/MetricTags.java

gosusnp reviewed (approved):
LGTM, I like the attribute grouping.

Left a couple of questions, also I am surprised that this change doesn't have a broader scope or are just we not using tags overall?

davinchia commented on airbyte-metrics/metrics-lib/src/main/java/io/airbyte/metrics/lib/OssMetricsRegistry.java:
```diff
@@ -95,7 +95,19 @@ public enum OssMetricsRegistry implements MetricsRegistry {
-      "overall job runtime - scheduling and execution for all attempts - for jobs that reach terminal states in the last hour. tagged by terminal states.");
+      "overall job runtime - scheduling and execution for all attempts - for jobs that reach terminal states in the last hour. tagged by terminal states."),
+
+  TEMPORAL_WORKFLOW_RESTART_ATTEMPT(MetricEmittingApps.WORKER,
+      "temporal.workflow.restart.attempt",
+      "count of number of attempts to restart a workflow"),
+
+  TEMPORAL_WORKFLOW_RESTART_SUCCESS(MetricEmittingApps.WORKER,
+      "temporal.workflow.restart.success",
+      "count of number of successful restarts performed by the workflow connection manager."),
+
+  TEMPORAL_WORKFLOW_RESTART_FAILURE(MetricEmittingApps.WORKER,
+      "temporal.workflow.restart.failure",
+      "count of the number of workflow restart failures");
```
Thinking out loud: 
These are all related to the sync workflow a.k.a the child workflow. 

1) Do we want to also track the connection manager workflow runs success/failure? I think it is and would love thoughts here.
2) I'm thinking we can make clearer what workflow these refer to, either via a name change, better descriptions, different tags etc. As is, I think this refers to all workflows when it only refers to temporal sync workflows.

jdpgrailsdev replied:
@davinchia I think that is the logical evolution of this (track syncs AND restarts).  The question becomes whether or not the metric should just be something like `temporal.workflow.(attempt|success|failure)` and have an attribute/tag of `type` (sync|restart value) or if the type should be part of the metric name.  I suppose the basic question to ask ourselves is whether or not we would want to see ALL workflows and/or be able to filter by type or if we will always just filter by type on a dashboard/alert.  We should answer this now to get the names correct, but we don't have to add the metrics for tracking the syncs as part of this PR -- it should probably be a separate PR in my opinion just so that it gets the proper focus.

benmoriceau reviewed (commented):
I think that I ddin't understand why we need a Tuple compare to just a list of tags.

davinchia reviewed (commented):
Thanks @jdpgrailsdev.

Some comments on better describing how the new attributes are meant to be used.

Two bigger notes:
1) We currently have the MetricTags class to keep track of tags similar to how the MetricRegistries help give some method to the madness. How do you think we should incorporate this with the new Metric Attributes?
2) Do we want to also track ConnectionManagerWorkflow success/failures?

jdpgrailsdev commented:
> I think that I ddin't understand why we need a Tuple compare to just a list of tags.

@benmoriceau The simple answer is that the OTEL protocol treats tags/attributes different than the StatsD protocol.  OTEL represents them as key/value pairs.  While it is true that we could parse the colon concatenated value, it seemed cleaner to represent it as it is (a key and value) and let the client generate the tag/attribute format instead of having all that parsing logic to break it back apart in some cases.

jdpgrailsdev commented:
@davinchia @benmoriceau @alovew @gosusnp 

I have addressed the PR feedback, minus the sync vs reset metrics (more on that in a bit).  While addressing the feedback, I noticed that the `MetricTags` class was concatenating the key/value into a colon-delimited string.  I missed this the first time.  I have updated the PR to remove that logic/move it to the StatsD client, as it is the only client that needs the attribute to be in that format.  This makes more sense to me, as the logic to concatenate is StatsD specific and not something used by the OTEL API.

As for @davinchia's comment about tracking the general syncs too, I agree that we should add those.  The question in my mind at this moment is whether or not we do that as part of this PR.  My feeling is that we decide on the naming convention and update this PR to reflect that (see my comment above in response to @davinchia 's original comment) and get that into this PR.  Then, we can do a separate PR to add the metrics for the syncs, so it doesn't get lost/overlooked in this PR.  However, I am open to just doing it as part of this PR if people feel that is easier.

jdpgrailsdev commented:
@davinchia I have updated the PR to generally record success/failures for workflows.  I attempted to record the type of the workflow (the class has a `isResetJob` method that looks at the state stored in the manager), but that caused tests to fail and I wasn't sure why.  

@benmoriceau is there another way in the `ConnectionManagerWorkflowImpl` that we can detect the type of the workflow (sync vs reset)?  Even if it involves passing the data in to the run method as part of/a change to the `ConnectionUpdaterInput` that would be something we could consider in this PR.  While I can't think of very many cases where we want to know the difference for the metrics, if having it is cheap enough, it may save us some time when debugging an issue.

alovew commented on airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:
```diff
@@ -123,10 +132,13 @@ public class ConnectionManagerWorkflowImpl implements ConnectionManagerWorkflow
 
   private UUID connectionId;
 
+  private final MetricClient metricClient = MetricClientFactory.getMetricClient();
```
this is where I'm not sure whether you'll have a metric client available. the metric client has to be initialized using MetricClient.initialize and the correct env variables need to be present where it's initialized.

jdpgrailsdev replied:
Ideally, that is something that will be handled by the framework.  I'll check on where we call initialize.  Either way, the code will just return the no-op client if it hasn't been initialized to avoid an NPE, so I think we are at least safe.

jdpgrailsdev replied:
@alovew I doubled checked and the `WorkerApp` calls initialize as the first thing it does.  Ultimately, this will be done by Micronaut but for now, it looks like everything is already set up for the Worker app.

davinchia commented on airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:
```diff
@@ -135,6 +147,7 @@ public void run(final ConnectionUpdaterInput connectionUpdaterInput) throws Retr
         // When a scope is cancelled temporal will throw a CanceledFailure as you can see here:
         // https://github.com/temporalio/sdk-java/blob/master/temporal-sdk/src/main/java/io/temporal/workflow/CancellationScope.java#L72
         // The naming is very misleading, it is not a failure but the expected behavior...
+        recordFailureCountMetric(connectionUpdaterInput, FailureCause.CANCELED);
```
I'm curious when did cancel would ever be triggered since I'd think any cancel is caught by the sync child workflow. @benmoriceau would you know?

davinchia reviewed (approved):
Thanks @jdpgrailsdev. Apologies I didn't get a chance to review this till now.

Some minor corrections/suggestions. Meat of this looks good so nothing here is blocking.

jdpgrailsdev commented on airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/ConnectionManagerWorkflowImpl.java:
```diff
@@ -160,7 +173,7 @@ public void run(final ConnectionUpdaterInput connectionUpdaterInput) throws Retr
     } catch (final Exception e) {
       log.error("The connection update workflow has failed, will create a new attempt.", e);
-      reportFailure(connectionUpdaterInput, null);
+      reportFailure(connectionUpdaterInput, null, FailureCause.UNKNOWN);
```
@davinchia I wasn't sure why this was calling `reportFailure` either, but I figured that I would not remove/change too much code just to get the metrics in place.  Once this has been merged, we can see how often (and why) this is happening to determine what should be done in this catch.  I do agree with your assessment:  this is catching any exception raised while handling the result of the workflow (e.g. lines 153-172 ish).
